dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec40/identitystore/InMemoryIdentityStoreDefinitionWrapper.java

```diff
@@ -0,0 +1,371 @@
+            }
+
+            Tr.warning(tc, "JAKARTASEC_WARNING_IDSTORE_CONFIG", new Object[] { "useFor/useForExpression", JakartaSec40Constants.SPEC_DEFAULT_USEFOR });
+
+            Set<ValidationType> values = new HashSet<ValidationType>();
+            values.addAll(Set.of(JakartaSec40Constants.SPEC_DEFAULT_USEFOR));
+
+            return values;
+        }
+    }
+
+    /**
+     * Get the priority for the {@link IdentityStore}.
+     *
+     * @return The priority.
+     *
+     * @see InMemoryIdentityStoreDefinition#priority()
+     * @see InMemoryIdentityStoreDefinition#priorityExpression()
+     */
+    int getPriority() {
+        return (this.priority != null) ? this.priority : evaluatePriority(false);
+    }
+
+    /**
+     * Get the useFor for the {@link IdentityStore}.
+     *
+     * @return The useFor.
+     *
+     * @see InMemoryIdentityStoreDefinition#useFor()
+     * @see InMemoryIdentityStoreDefinition#useForExpression()
+     */
+    Set<ValidationType> getUseFor() {
+        return (this.useFor != null) ? this.useFor : evaluateUseFor(false);
+    }
+
+    /**
+     * Get a specific credential from the credentials map given a caller name.
+     *
+     * @param callerName is the caller name key to the map.
+     * @return the CredentialValue, or null if the caller name is invalid.
+     */
+    public CredentialValue getCredential(String callerName) {
+        return credentials.get(callerName);
+    }
+
+    /***********************************************************************
+     * Helper classes
+     ***********************************************************************/
+
+    private static class ELHelperWrapper extends ELHelper {
+
+        /**
+         * Required as protected in base class.
+         */
+        @Override
+        public Set<ValidationType> processUseFor(String useForExpression, ValidationType[] useFor, boolean immediateOnly) {
+            return super.processUseFor(useForExpression, useFor, immediateOnly);
+        }
+
+        /**
+         * Evaluate possible string EL expressions
+         */
+        @Override
+        public String processString(String name, @Sensitive String expression, boolean immediateOnly, boolean mask) throws IllegalArgumentException {
+            return super.processString(name, expression, immediateOnly, mask);
+        }
+    }
+
+    /**
+     * This class models the Credentials interface, but is specific for use within
+     * the identitystore package, modelling the data, and providing minimal APIs for
+     * external use.
+     */
+
+    static final class CredentialValue extends CredentialPassword {
+
+        private final String callerName;
+        private final String[] groups;
+
+        CredentialValue(final String callerName, final @Sensitive String password, final String[] groups) {
+            super(password);
+            this.callerName = callerName;
+            this.groups = Arrays.copyOf(groups, groups.length);
```

Comment:
If groups can be null as you have in the `getGroups` method, this will cause a NullPointerException.  We seem to copy groups over and over in here and in getGroups.  Is it exposed externally?

Reply:
Yes, `getGroups` is exposed, but within the same package .  So one `copyOf` for the way in to disallow external changes after creation, and the other for the way out to disallow external tampering.  Maybe a bit excessive?  But please suggest alternatives.

Reply:
Ok that is fine.  `getGroups()` handles the case that the `groups` variable is `null`, but this method does not which was the main question I had.
```suggestion
            this.groups = groups == null ? null : Arrays.copyOf(groups, groups.length);
```
That is what I am suggesting.

Reply:
Okay, yep, null pointer exception if parameter `groups` is null.

Done.